### ISSUE TITLE
Questions about tsp.cpp implementation details

### ISSUE THREAD

**Issue by kevin031060** [DIMESTeam/DIMES#8]
After reading and running your paper and code, I found that the training process is both fast and effective, which is truly a great job. I am currently trying to apply your method to a new paper I'm working on, but I have some questions regarding the tsp.cpp implementation:
1. My understanding is that the sampling method relies on random sampling based on par (heatmap). However, the sample() function in tsp.cpp utilizes segment_max_csr to select the maximum without performing random sampling based on probability. Also, I am unclear about the role of sample_cands(). It seems to update v by choosing the closest node as the next node, which doesn't appear to sample the next candidate randomly by the heatmap.

2. It seems that the only difference between TspGreedySampler and TspSoftmaxSampler lies in the init() function. While TspGreedySampler generates non-repeating initial nodes of the TSP using randperm, TspSoftmaxSampler allows for repeating initial nodes with randint. It is not like the common way of greedy and sampling in NCO.

3. I am curious about the role of sample_size. Why does tsp.cpp process par sample_size times? Is it to address symmetry in the problem?

I would greatly appreciate your insights on these points. Understanding these aspects will significantly help me in adapting your approach to my current project.

Thank you for your time and consideration.

Best regards,
kevin

**Comment by q-rz** [comment]
Hi Kevin,

Thanks for your interest in our work! My clarifications to your questions are as follows.

1. We use `segment_max_csr` because we use the **Gumbel-max trick** (see Section 2 in [this paper](https://arxiv.org/pdf/1611.01144.pdf)): sampling from a categorical distribution $\textrm{Categorical}(p_1,\dots,p_n)$ is equivalent to finding $\textrm{argmax}_{i=1}^n(g_i+\log p_i)$ where $g_1,\dots,g_n\sim\textrm{Gumbel}(0,1)$. The Gumbels are generated on [line 158](https://github.com/DIMESTeam/DIMES/blob/main/TSP/torch_sampling/csrc/tsp.cpp#L158C28-L158C81): `-torch::empty_like(this->par_e).exponential_().log()`.
  Regarding the `sample_cands` function, it is only for a corner case: in the KNN graph, it can happen that the $k$ candidate next nodes have all been visited before. In this case, since we have to choose a node from the remaining non-adjacent nodes, which are not included in the heatmap, we choose the closest node as the next node. In normal cases, we do not use the `sample_cands` function. (I just checked my paper and found that I forgot to mention this in Appendix C.1... Sorry for the confusion.)
2. The major difference is that `TspGreedySampler` does not use the Gumbel-max trick while `TspSoftmaxSampler` uses the Gumbel-max trick. Thus, `TspGreedySampler` is indeed "greedy" w.r.t. the heatmap while `TspSoftmaxSampler` follows the categorical distribution defined by the heatmap. 
3. `sample_size` is the number of solutions we want to obtain, which is just a hyperparameter. I do not quite understand your question "Why does tsp.cpp process par sample_size times?". Could you elaborate more on it?

If you have any further questions, please feel free to let me know.

Warm regards,
Ruizhong

**Comment by kevin031060** [comment]
Very thanks for your response, I didn't know the Gumbel max trick before. Now I understand the code. However, why we need to use the Gumbel max trick, which in my understanding is to make the sampling process differentiate. The REINFORCE algorithm uses the outcome (reward) of an action (selected through sampling or argmax) to adjust the policy  in a direction that increases expected reward. This is achieved without needing to differentiate through the sampling or argmax operation itself. 

Moreover, the sampling or greedy process is executed **sample_size** times. But in previous papers, for processing an instance in training, the sampling process is typically performed only once to acquire the reward and log(pi).

**Comment by q-rz** [comment]
In fact, I used the Gumbel max trick just because I think it has better numerical stability than explicitly computing the probabilities... (I did not experimentally verify this hypothesis, though.) By the way, the Gumbel max trick does not make it differentiable; only the Gumbel softmax trick enables differentiability.

Regarding `sample_size`, we use a small number of samples to estimate the gradient because it gives a more accurate gradient estimate without significantly increasing the running time. It is like mini-batch SGD v.s. SGD.
